source/Mlos.Python/mlos/Optimizers/RegressionModels/RegressionEnhancedRandomForestModel.py

```diff
@@ -387,214 +310,133 @@ def _execute_grid_search_for_random_forest_regressor_model(self, x_filtered_to_d
         self.model_config.perform_initial_random_forest_hyper_parameter_search = False
 
     @trace()
-    def predict(self, feature_values_pandas_frame, include_only_valid_rows=True):
-        check_is_fitted(self)
-
-        x_filtered_to_detected_features = self.transform_x(feature_values_pandas_frame)
-
-        base_predicted = self.base_regressor_.predict(x_filtered_to_detected_features)
-        residual_predicted = self.random_forest_regressor_.predict(x_filtered_to_detected_features)
-        y_predicted = base_predicted + residual_predicted
+    def predict(
+            self,
+            feature_values_pandas_frame: pd.DataFrame,
+            include_only_valid_rows: bool = True
+    ) -> Prediction:
 
-        # dataframe column shortcuts
+        # Prediction dataframe column shortcuts
+        is_valid_input_col = Prediction.LegalColumnNames.IS_VALID_INPUT.value
         predicted_value_col = Prediction.LegalColumnNames.PREDICTED_VALUE.value
         predicted_value_var_col = Prediction.LegalColumnNames.PREDICTED_VALUE_VARIANCE.value
-        predicted_value_dof_col = Prediction.LegalColumnNames.DEGREES_OF_FREEDOM.value
+        dof_col = Prediction.LegalColumnNames.PREDICTED_VALUE_DEGREES_OF_FREEDOM.value
+
+        valid_rows_index = None
+        model_design_matrix: np.ndarray = None
+        model_design_matrix_dataframe: pd.DataFrame = None
+        if self.trained:
+            if self.x_is_design_matrix:
+                model_design_matrix = feature_values_pandas_frame.to_numpy()
+                model_design_matrix_dataframe = feature_values_pandas_frame
+            else:
+                features_df = self.input_space.project_dataframe(feature_values_pandas_frame, in_place=False)
+                (model_design_matrix, new_column_names) = self._transform_x(features_df)
+                model_design_matrix_dataframe = pd.DataFrame(model_design_matrix, columns=new_column_names)
+            valid_rows_index = feature_values_pandas_frame.index
 
         # initialize return predictions
-        predictions = RegressionEnhancedRandomForestRegressionModelPrediction(objective_name=self.output_dimension_names[0])
-        prediction_df = predictions.get_dataframe()
+        predictions = Prediction(
+            objective_name=self.target_dimension_names[0],
+            predictor_outputs=self._PREDICTOR_OUTPUT_COLUMNS,
+            dataframe_index=valid_rows_index
+        )
+        prediction_dataframe = predictions.get_dataframe()
+
+        if valid_rows_index is not None and not valid_rows_index.empty:
+            prediction_dataframe[is_valid_input_col] = True
+
+            base_predictions_dataframe = self.base_regressor_.predict(model_design_matrix_dataframe).get_dataframe()

```

Comment:
At any point do we validate that this design matrix, belongs to the base regressors feature space?  I mean, it seems like it should, but at least for debugging or somewhere in tests, an assert would be good. Even if only on column names being what the base regressor expects.

Reply:
Maybe assert that the design_matrix column names are made up of polynomial terms and one hot encoding column names derived from the RERF input_space features?  I'm not sure that failing such an assert would mean anything.  For example, if Ed+1 changes the transform_x() method and destroys the feature names, but correctly captures the design, then asserting about column names may cause pause, but not anything more meaningful.  What we want to be certain of is that the base regressors don't encounter categorical dimensions and then invoke their OHE adapter because the OHE adapter output_space is combined with the polynomial adapter output_space is a very specific manner.  Constructing the design_matrix_hypergrid ensures the base regressor is initialized with the expected input_space needed to correspond to the exploded feature space.